### PR TITLE
aarch64: add necessary kvm ioctls

### PR DESCRIPTION
**Pull request by dianpopa**
`cargo test` is functional on aarch64 too.
Fixes https://github.com/firecracker-microvm/firecracker/issues/1040.

Signed-off-by: Diana Popa <dpopa@amazon.com>

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
Before merging this PR I would like to setup a CI for ARM as well to make sure that future PRs don't break the ARM build.

**Comment by dianpopa** [comment]
> Before merging this PR I would like to setup a CI for ARM as well to make sure that future PRs don't break the ARM build.

Seems like a good idea. However, until then, any feedback is welcome.

File: src/ioctls/device.rs
Comment (andreeaflorescu) [inline src/ioctls/device.rs:1]: The current structure of kvm-ioctls is to have  3 sub-modules in the `ioctls` module that correspond to the KVM ioctls: system, vm and vcpu. When helper structures are required we can do one of two things:
- if they're used by more than one sub-module (system, vm or vcpu) the definition can lie in ioctls/mod.rs
- it they're used by only one sub-module than they should be added to the sub-module that uses them.

I believe `DeviceFd` is used only by `VmFd` and if this is not going to change in the future I say we place this code there rather than creating a new file so we can keep the initial structure.

File: src/ioctls/device.rs
Comment (andreeaflorescu) [inline src/ioctls/device.rs:22]: Please add a short description of the ioctl: 
```
/// Sets a specified piece of device configuration and/or state.
///
/// See the documentation for `KVM_SET_DEVICE_ATTR`.
```

File: src/ioctls/device.rs
Comment (andreeaflorescu) [inline src/ioctls/device.rs:63]: nit: can you import the structures and constants that you need from the kvm_bindings instead of using them with kvm_bindings::KVM_CREATE_DEVICE_TEST? The code is a bit hard to read.

File: src/ioctls/device.rs
Comment (andreeaflorescu) [inline src/ioctls/device.rs]: This is also expected to fail on ARM. Please update the comment.

File: src/ioctls/vcpu.rs
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs:305]: Please format the doc comment with summary and description. In the description also add the KVM ioctl that the method is using (KVM_ARM_VCPU_INIT). Since this ioctl has a parameter also add the `# Arguments` section.

Same for `set_one_reg`.

File: src/ioctls/vcpu.rs
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs]: s/becuase/because

File: src/ioctls/vcpu.rs
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs:807]: Can you add a comment here regarding what are you trying to test and what is the `PSTATE_REG_ID`. This would be useful for people looking at the tests in the future.

File: src/ioctls/vcpu.rs
Comment (andreeaflorescu) [inline src/ioctls/vcpu.rs]: nit: We can change this expect into unwrap to get rid of the "bla".

File: src/ioctls/vm.rs
Comment (andreeaflorescu) [inline src/ioctls/vm.rs:270]: We should try to name methods as close as possible to the KVM ioctls that they provide wrappers for. Usually the names of all methods strip the `KVM_` prefix and use lower case. For this one it would be arm_preferred_target.

File: src/ioctls/vm.rs
Comment (andreeaflorescu) [inline src/ioctls/vm.rs:401]: With this change the following asserts fail for different reasons on arm and x86. I think it is worth mentioning why we expect the asserts to fail on each platform.

**Review by andreeaflorescu — COMMENTED** [review]
General comments:
- Please update the coverage to 91.3. This is x86 specific unfortunately, but I added an issue to take care of it for ARM as well:  #9 .The coverage value is stored in tests/coverage.
- All doc comments (larger than one line) should be split in summary and description. Also, for methods that wrap KVM ioctls add a comment with the name of the ioctl that they provide wrappers for. If the methods have arguments, please add the `# Arguments` section to the comment.

**Comment by andreeaflorescu** [comment]
@dianpopa I added a temporary fix for running cargo build with gnu/musl, and cargo test with gnu/musl on arm as well.

We still need to work on adding a coverage test for aarch64, but other than that we should be good to merge the PR once the tests pass.

The musl unit tests seem to fail for a reason that is unrelated to your PR. Do you happen to know why is there a linking problem? You can check the logs here: https://buildkite.com/rust-vmm/kvm-ioctls-ci/builds/108#55d3065d-1563-4a50-a0c9-2edd7c3f6afa

File: src/ioctls/device.rs
Comment (dianpopa) [reply src/ioctls/device.rs:1]: DeviceFd is not just a helper structure. KVM_CREATE_DEVICE (just like KVM_CREATE_VPCU and KVM_CREATE_VM) will give back an fd on which later you can call other kvm ioctls. Take a look at 
```
4.80 /KVM_GET_DEVICE_ATTR
Type: device ioctl, vm ioctl, vcpu ioctl
```
from the api.txt documentation. So `DeviceFd` corresponds to another type of ioctl, device ioctl (apart from vcpu and vm).
"I believe DeviceFd is used only by VmFd" -> DeviceFd will be used by regs and the global interrupt controller when all arm64 functionality will be merged.
I actually considered putting the code inside VMFd, however in VmFd is a lot of code already and it made it easier to add unit tests and add documentation putting it in a separate file.
If you still think it makes more sense to put it with VmFd, let me know.

File: src/ioctls/vm.rs
Comment (dianpopa) [reply src/ioctls/vm.rs:270]: I find it ugly to put a cfg above this method that says it is arm specific and then name that method with an arm prefix. 
I think it is nice to keep that rule around but we can make exceptions when we feel there is a need to (see the register_irqfd function).

File: src/ioctls/device.rs
Comment (andreeaflorescu) [reply src/ioctls/device.rs:1]: Gotcha! Sounds good as it it. I checked the KVM API again, I missed the device ioctls entry before.

File: src/ioctls/mod.rs
Comment (andreeaflorescu) [inline src/ioctls/mod.rs:1]: nit: Adding the license is an unrelated change. Can you please add it in a new commit?

File: src/ioctls/device.rs
Comment (andreeaflorescu) [inline src/ioctls/device.rs:15]: The structure need to be publicly exported in src/lib.rs to be accessible outside of the crate.

Comment (dianpopa) [reply src/ioctls/device.rs:15]: Done, see lib.rs file.

File: src/ioctls/mod.rs
Comment (dianpopa) [reply src/ioctls/mod.rs:1]: Done, take a look.

**Review by andreeaflorescu — APPROVED** [review]

**Review by sameo — APPROVED** [review]

**Pull request merged**
